### ISSUE TITLE
Integrity test or the KATs fail, but cannot get "PROV_R_MODULE_INTEGRITY_FAILURE" error info

### ISSUE THREAD

**Issue by yangjinlin01** [openssl/openssl#20614]

**Comment by paulidale** [comment]
What output are you getting on failure?  The module integrity failure message is only for when the module's checksum changes, it wouldn't show for other selftest errors.

In the build tree I can run:
```sh
$  ./util/wrap.pl apps/openssl fipsinstall  -out fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
HMAC : (KAT_Integrity) : Pass
HMAC : (Module_Integrity) : Pass
SHA1 : (KAT_Digest) : Pass
SHA2 : (KAT_Digest) : Pass
SHA3 : (KAT_Digest) : Pass
AES_GCM : (KAT_Cipher) : Pass
AES_ECB_Decrypt : (KAT_Cipher) : Pass
RNG : (Continuous_RNG_Test) : Pass
RSA : (KAT_Signature) : Pass
DSA : (KAT_Signature) : Pass
TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
TLS13_KDF_EXPAND : (KAT_KDF) : Pass
TLS12_PRF : (KAT_KDF) : Pass
PBKDF2 : (KAT_KDF) : Pass
SSHKDF : (KAT_KDF) : Pass
KBKDF : (KAT_KDF) : Pass
HKDF : (KAT_KDF) : Pass
SSKDF : (KAT_KDF) : Pass
X963KDF : (KAT_KDF) : Pass
X942KDF : (KAT_KDF) : Pass
HASH : (DRBG) : Pass
CTR : (DRBG) : Pass
HMAC : (DRBG) : Pass
DH : (KAT_KA) : Pass
RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
	name:     	OpenSSL FIPS Provider
	version:  	3.2.0
	build:    	3.2.0-dev
INSTALL PASSED
```
Then check that everything is happy:
```sh
$ ./util/wrap.pl apps/openssl fipsinstall -verify -in fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
VERIFY PASSED
```
Finally, modify something on `module-mac` line in the `fips.cnf` file to get the integrity failure:
```sh
$ ./util/wrap.pl apps/openssl fipsinstall -verify -in fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
Module integrity mismatch
VERIFY FAILED
```

My guess is you've got your paths set incorrectly and things are not being found properly.

**Comment by slontis** [comment]
It could also happen if you change the source and dont update the fips config file checksum.

**Comment by yangjinlin01** [comment]
I change module-mac value in fipsmodule.cnf, and test fips status, I expect it print out “

PROV_R_MODULE_INTEGRITY_FAILURE” ,but it only print out  “invalid status”.

After debugging, I find that  module_run return 1, and all error info clear by ERR_pop_to_mark.





发件人: Pauli ***@***.***
发送时间: 2023年3月28日 5:41
收件人: openssl/openssl ***@***.***>
抄送: Devin Yang (杨金林)-云数据中心集团 ***@***.***>; Author ***@***.***>
主题: Re: [openssl/openssl] Integrity test or the KATs fail,but cannot get "PROV_R_MODULE_INTEGRITY_FAILURE" error info (Issue #20614)



What output are you getting on failure? The module integrity failure message is only for when the module's checksum changes, it wouldn't show for other selftest errors.

In the build tree I can run:

$  ./util/wrap.pl apps/openssl fipsinstall  -out fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
HMAC : (KAT_Integrity) : Pass
HMAC : (Module_Integrity) : Pass
SHA1 : (KAT_Digest) : Pass
SHA2 : (KAT_Digest) : Pass
SHA3 : (KAT_Digest) : Pass
AES_GCM : (KAT_Cipher) : Pass
AES_ECB_Decrypt : (KAT_Cipher) : Pass
RNG : (Continuous_RNG_Test) : Pass
RSA : (KAT_Signature) : Pass
DSA : (KAT_Signature) : Pass
TLS13_KDF_EXTRACT : (KAT_KDF) : Pass
TLS13_KDF_EXPAND : (KAT_KDF) : Pass
TLS12_PRF : (KAT_KDF) : Pass
PBKDF2 : (KAT_KDF) : Pass
SSHKDF : (KAT_KDF) : Pass
KBKDF : (KAT_KDF) : Pass
HKDF : (KAT_KDF) : Pass
SSKDF : (KAT_KDF) : Pass
X963KDF : (KAT_KDF) : Pass
X942KDF : (KAT_KDF) : Pass
HASH : (DRBG) : Pass
CTR : (DRBG) : Pass
HMAC : (DRBG) : Pass
DH : (KAT_KA) : Pass
RSA_Encrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
RSA_Decrypt : (KAT_AsymmetricCipher) : Pass
        name:          OpenSSL FIPS Provider
        version:       3.2.0
        build:         3.2.0-dev
INSTALL PASSED

Then check that everything is happy:

$ ./util/wrap.pl apps/openssl fipsinstall -verify -in fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
VERIFY PASSED

Finally, modify something on module-mac line in the fips.cnf file to get the integrity failure:

$ ./util/wrap.pl apps/openssl fipsinstall -verify -in fips.cnf -module providers/fips.so -provider_name fips -macopt digest:SHA256 -section_name fips_sect
Module integrity mismatch
VERIFY FAILED

My guess is you've got your paths set incorrectly and things are not being found properly.

**Comment by slontis** [comment]
Debug the function SELF_TEST_post() to see where that fails..

**Comment by yangjinlin01** [comment]
> Debug the function SELF_TEST_post() to see where that fails..
**this is my test code，I have enable fips in openssl.cnf, and change  module-mac value to a wrong value in fipsmodule.cnf, i expect"PROV_R_MODULE_INTEGRITY_FAILURE" error info, but
it only print：
error data:error:1C8000D4:Provider routines::invalid state
error data:error:1C8000D8:Provider routines::self test post failure
error data:error:078C0105:common libcrypto routines::init fail .
when i debug, i found SELF_TEST_post is called twice.First call i can find the right error info "PROV_R_MODULE_INTEGRITY_FAILURE", but when module_init function return 1,  the function ERR_pop_to_mark clear all error info;second call ,loclstate variale has been changed ,and  it will ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_STATE).
My compile command: gcc -g test.c -o test -lcrypto**
`#include <stdio.h>
#include <stdlib.h>
#include <openssl/provider.h>
#include <openssl/evp.h>
#include <openssl/core.h>
#include <openssl/err.h>
void print_errors()
{
    int flag = 0, line = 0;
    char *data = NULL, *file = NULL;
    unsigned long code = 0;
    code = ERR_get_error_line_data(&file, &line, &data, &flag);
    while (code) {
        printf("ERROR code :%lu in %s line %d.\n", code,.file, line);
        //if (data &&(flag & ERR_TXT_STRING))
            printf("error data:%s\n", data);
            code = ERR_get_error_line_data(&file, &line, &data, &flag);

    }
}
int main(void){
    const OSSL_PROVIDER *prov = OSSL_PROVIDER_load(NULL, "fips");
    if (prov == NULL) {
        print_errors();
        return -1;
    }
    OSSL_PARAM  param[5];
    char *name, *version, *buildinfo;
    int status;
    param[0] = OSSL_PARAM_construct_utf8_ptr("name", &name, 0);
    param[1] = OSSL_PARAM_construct_utf8_ptr("version", &version, 0);
    param[2] = OSSL_PARAM_construct_utf8_ptr("buildinfo", &buildinfo, 0);
    param[3] = OSSL_PARAM_construct_int("status", &status);
    param[4] = OSSL_PARAM_construct_end();

    if (!OSSL_PROVIDER_get_params(prov, param)){
        printf("ERROR\n");
    }
    printf("Name:%s\n", name);
    printf("Version:%s\n", version);
    printf("Build info: %s\n", buildinfo);
    if (status == 1) {
        printf("Status: activate\n");
    } else {
        printf("Status: Non-activate\n");
    }

    return 0;
}`

**Comment by slontis** [comment]
If you need this information then use OSSL_LIB_CTX_load_config() instead...

To try to explain what is happening..
When you just use OSSL_PROVIDER_load() the FIPS module requires the fips config in order to work.. So a call to the load internally uses the openssl-config file to determine what modules to run.. When it generically loads like this the order that things happens in causes the self tests to run before other things have been setup. This is why you end up with 2 calls.. The first one loads the fips provider but has to return generically. The FIPS provider then sets itself into an error state, so any calls to it after this point will return an error status. So that is why the second call returns the error status.

You could also use the OSSL_SELF_TEST_set_callback().  See doc/man7/OSSL_PROVIDER-FIPS.pod and doc/man3/OSSL_SELF_TEST_set_callback.pod for more info. This functions runs each time a self test happens with info about the test that you can then log (Which is the output above that fipsinstall uses).

**Comment by slontis** [comment]
@paulidale Because of the changes in 3.1, I wonder if this now cause the self tests to run twice in this scenario?

**Comment by yangjinlin01** [comment]
@slontis  I find the function "provider_conf_load" always return 1, and the comment in the openssl source code is
   /*
     * Even if ok is 0, we still return success. Failure to load a provider is
     * not fatal. We want to continue to load the rest of the config file.
     */   ,can I change it to return ok

**Comment by paulidale** [comment]
> @paulidale Because of the changes in 3.1, I wonder if this now causes the self tests to run twice in this scenario? (Note this comment is not directed towards you @yangjinlin01)

It seems like it does.  Not ideal but probably no way to avoid it -- the provider is being loaded twice after all.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
